File: VB/Q245342/MySpinEdit.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Windows.Forms
Imports System.ComponentModel
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator

Namespace DXSample
	Public Class MySpinEdit
		Inherits SpinEdit
		Shared Sub New()
			RepositoryItemMySpinEdit.RegisterMySpinEdit()
		End Sub
		Public Sub New()
			MyBase.New()
			fValueUpdateTimer = New Timer()
			fValueUpdateTimer.Interval = 200
			AddHandler ValueUpdateTimer.Tick, Function(sender, e) AnonymousMethod1(sender, e)
			ValueUpdateStep = 0D
		End Sub
		
		Private Function AnonymousMethod1(ByVal sender As Object, ByVal e As EventArgs) As Boolean
			Value += ValueUpdateStep
			Return True
		End Function



		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemMySpinEdit.MySpinEditName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemMySpinEdit
			Get
				Return CType(MyBase.Properties, RepositoryItemMySpinEdit)
			End Get
		End Property

		Private fValueUpdateTimer As Timer
		Private ReadOnly Property ValueUpdateTimer() As Timer
			Get
				Return fValueUpdateTimer
			End Get
		End Property
		Private fValueUpdateStep As Decimal
		Private Property ValueUpdateStep() As Decimal
			Get
				Return fValueUpdateStep
			End Get
			Set(ByVal value As Decimal)
				fValueUpdateStep = value
			End Set
		End Property

		Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)
			If e.Button = MouseButtons.Left Then
				ValueUpdateStep = If(e.Location.Y > Height / 2, -1D, 1D)
				If ViewInfo.CalcHitInfo(e.Location).HitTest = EditHitTest.Button Then
					ValueUpdateTimer.Start()
				End If
			End If
		End Sub

		Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			MyBase.OnMouseUp(e)
			If ValueUpdateTimer.Enabled Then
				ValueUpdateTimer.Stop()
			End If
		End Sub
	End Class

	<UserRepositoryItem("RegisterMySpinEdit")> _
	Public Class RepositoryItemMySpinEdit
		Inherits RepositoryItemSpinEdit
		Shared Sub New()
			RegisterMySpinEdit()
		End Sub
		Public Sub New()
			MyBase.New()
		End Sub

		Friend Const MySpinEditName As String = "MySpinEdit"

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return MySpinEditName
			End Get
		End Property

		Public Shared Sub RegisterMySpinEdit()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(MySpinEditName, GetType(MySpinEdit), GetType(RepositoryItemMySpinEdit), GetType(BaseSpinEditViewInfo), New ButtonEditPainter(), True))
		End Sub
	End Class
End Namespace

File: CS/Q245342/MySpinEdit.cs
using System;
using System.Windows.Forms;
using System.ComponentModel;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;

namespace DXSample {
    public class MySpinEdit :SpinEdit {
        static MySpinEdit() { RepositoryItemMySpinEdit.RegisterMySpinEdit(); }
        public MySpinEdit() : base() { 
            fValueUpdateTimer = new Timer();
            fValueUpdateTimer.Interval = 200;
            ValueUpdateTimer.Tick += delegate(object sender, EventArgs e) { Value += ValueUpdateStep; };
            ValueUpdateStep = 0m;
        }



        public override string EditorTypeName { get { return RepositoryItemMySpinEdit.MySpinEditName; } }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMySpinEdit Properties { 
            get { return (RepositoryItemMySpinEdit)base.Properties; } 
        }

        private Timer fValueUpdateTimer;
        private Timer ValueUpdateTimer { get { return fValueUpdateTimer; } }
        private decimal fValueUpdateStep;
        private decimal ValueUpdateStep { get { return fValueUpdateStep; } set { fValueUpdateStep = value; } }

        protected override void OnMouseMove(MouseEventArgs e) {
            base.OnMouseMove(e);
            if (e.Button == MouseButtons.Left) {
                ValueUpdateStep = e.Location.Y > Height / 2 ? -1m : 1m;
                if (ViewInfo.CalcHitInfo(e.Location).HitTest == EditHitTest.Button) 
                    ValueUpdateTimer.Start();
            }
        }

        protected override void OnMouseUp(MouseEventArgs e) {
            base.OnMouseUp(e);
            if (ValueUpdateTimer.Enabled) ValueUpdateTimer.Stop();
        }
    }

    [UserRepositoryItem("RegisterMySpinEdit")]
    public class RepositoryItemMySpinEdit :RepositoryItemSpinEdit {
        static RepositoryItemMySpinEdit() { RegisterMySpinEdit(); }
        public RepositoryItemMySpinEdit() : base() { }

        internal const string MySpinEditName = "MySpinEdit";

        public override string  EditorTypeName { get { return MySpinEditName; } }

        public static void RegisterMySpinEdit() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(MySpinEditName, typeof(MySpinEdit),
                typeof(RepositoryItemMySpinEdit), typeof(BaseSpinEditViewInfo), new ButtonEditPainter(), 
                true));
        }
    }
}